bpo-41100: add runtime checks for MACOSX_DEPLOYMENT_TARGET=10.10

lawrence-danna-apple opened the pull request:
In order to support a universal2 build, supporitng Mac OS 11 on arm64 and Mac OS on
x86_64 going back to 10.10, we need to add in runtime checks for functions that will
be detected as present by autoconf, because they are in the SDK, but which did not
exist  in Mac OS 10.10.    This fixes all the instances of -WWunguarded-availability-new
when building with MACOSX_DEPLOYMENT_TARGET=10.10


<!-- issue-number: [bpo-41100](https://bugs.python.org/issue41100) -->
https://bugs.python.org/issue41100
<!-- /issue-number -->

lawrence-danna-apple committed:
add runtime checks for MACOSX_DEPLOYMENT_TARGET=10.10

In order to support a universal2 build, supporitng Mac OS 11 on arm64 and Mac OS on
x86_64 going back to 10.10, we need to add in runtime checks for functions that will
be detected as present by autoconf, because they are in the SDK, but which did not
exist  in Mac OS 10.10.    This fixes all the instances of -WWunguarded-availability-new
when building with MACOSX_DEPLOYMENT_TARGET=10.10

lawrence-danna-apple committed:
fix test_getgroups() on Mac OS for users with large numbers of groups

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

ronaldoussoren commented on configure.ac:
```diff
@@ -3825,6 +3825,19 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
    AC_MSG_RESULT(yes)
 ])
 
+AC_MSG_CHECKING([to see if the compiler supports __builtin_available])
```
It is not necessary to test for this in a configure script, the following should work in a header file:

```
#ifdef __has_builtin
#if __has_builtin(__builtin_available)
#define HAVE_BUILTIN_AVAILABLE 1
#endif
#endif
```

lawrence-danna-apple replied:
ok

ronaldoussoren commented on Modules/posixmodule.c:
```diff
@@ -9850,60 +9906,65 @@ os_pwritev_impl(PyObject *module, int fd, PyObject *buffers, Py_off_t offset,
-    int async_err = 0;
-    struct iovec *iov;
-    Py_buffer *buf;
+    if (HAVE_PWRITEV_RUNTIME) {
```
same as in os_preadv.

lawrence-danna-apple replied:
ok

ronaldoussoren commented on Modules/posixmodule.c:
```diff
@@ -9255,59 +9306,64 @@ os_preadv_impl(PyObject *module, int fd, PyObject *buffers, Py_off_t offset,
-    int async_err = 0;
-    struct iovec *iov;
-    Py_buffer *buf;
+    if (HAVE_PREADV_RUNTIME) {
```
I'd just ignore the compiler warning here, as this function will be removed from the module when HAVE_PREADV_RUNTIME is false.

lawrence-danna-apple replied:
ok

ronaldoussoren commented on Modules/posixmodule.c:
```diff
@@ -79,6 +79,28 @@ corresponding Unix manual entries for more information on calls.");
+#define HAVE_FUTIMENS_RUNTIME  __builtin_available(macos 10.13, ios 11, tvos 11, watchos 4, *)
+#define HAVE_PREADV_RUNTIME    __builtin_available(macos 10.16, ios 14, tvos 14, watchos 7, *)
+#define HAVE_PWRITEV_RUNTIME    __builtin_available(macos 10.16, ios 14, tvos 14, watchos 7, *)
+#define HAVE_POSIX_SPAWN_SETSID_RUNTIME __builtin_available(macos 10.15, ios 13, tvos 13, watchos 6, *)
```
Is this (the introduction of POSIX_SPAWN_SETSID) documented somewhere?

lawrence-danna-apple replied:
I couldn't find any proper documentation, I got 10.15 by looking up which version of xnu introduced it.

https://opensource.apple.com/source/xnu/xnu-4903.270.47/bsd/sys/spawn.h.auto.html
https://opensource.apple.com/source/xnu/xnu-6153.11.26/bsd/sys/spawn.h.auto.html

cdavis5e replied:
It is 10.15. I know because I asked for it (cf. FB5361314 aka rdar://43247036).

ronaldoussoren reviewed (commented):
Just some comments at this time. The basic approach looks sane, but I haven't look at the changes in detail yet.

lawrence-danna-apple committed:
fixup

lawrence-danna-apple committed:
fixup

lawrence-danna-apple committed:
fixup runtime checks

lawrence-danna-apple committed:
simplify

These complications were needed on a previous build of Xcode but as of
XCode 12 beta 3 it works without them.
